### PR TITLE
[Sampling] Cluster-GCN and ShaDow-GNN DataLoader

### PR DESCRIPTION
**Pull request by BarclayII**
## Description
This is an implementation of two subgraph samplers: Cluster-GCN and ShaDow-GNN, which are representative for two kinds of subgraph samplers:
* One that iterates over the subgraphs without explicit notion of seed nodes/edges (e.g. Cluster-GCN, GraphSAINT).
* Otherwise (e.g. ShaDow-GNN, which expands the seed nodes' neighborhood much like neighbor sampler but returns an induced subgraph instead).

Basically, the experience for both samplers looks like follows:

```python
# Cluster-GCN
dataset = dgl.data.CoraFullDataset()
g = dataset[0]
sgiter = dgl.dataloading.ClusterGCNSubgraphIterator(g, 100, '.', refresh=True)
dataloader = dgl.dataloading.GraphDataLoader(sgiter, batch_size=4, num_workers=num_workers)
for sg in dataloader:
    train_on(sg)
```

```python
# ShaDow-GNN
dataset = dgl.data.CoraFullDataset()
g = dataset[0]
sampler = dgl.dataloading.ShaDowKHopSampler([3, 3])
dataloader = dgl.dataloading.NodeDataLoader(g, train_nids, sampler, batch_size=4, num_workers=num_workers)
for input_nodes, output_nodes, (subgraph,) in dataloader:
    train_on(sg)
```

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by VoVAllen** [comment]
Potentially we might want support loading subgraph from disk for ClusterGCNTraining. And also is it possible to use NodeDataloader inside the GraphDataloader loop?

**Comment by BarclayII** [comment]
> In future we might want to support loading subgraph from disk for ClusterGCNTraining.

This should be doable with a switch of the cache format.  Currently it's saving the partition assignment and I can switch it to saving the actual partitions.  I'll leave it to a later PR though.

> is it possible to use NodeDataloader inside the GraphDataloader loop?

This is tricky.  ClusterGCN does not have a notion of seed nodes so I think `NodeDataLoader` cannot be used here.  It might be possible if we extend `NodeDataLoader` to allow customized minibatch iteration but I'm not sure what is the best practice (could be possible using PyTorch's custom minibatch sampler).

File: python/dgl/dataloading/cluster_gcn.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        The path to the cache directory for storing the partition result.
+    refresh : bool
+        If True, recompute the partition.
+
+    Examples
+    --------
+    Assuming that you have a graph ``g``:
+
+    >>> sgiter = dgl.dataloading.ClusterGCNSubgraphIterator(g, 100, '.', refresh=True)
+    >>> dataloader = dgl.dataloading.GraphDataLoader(sgiter, batch_size=4, num_workers=num_workers)
```
Comment (mufeili) [inline python/dgl/dataloading/cluster_gcn.py:38]: Give a specific value for `num_workers` for consistency.

**Comment by mufeili** [comment]
Update the `.rst` files for documentation.

File: python/dgl/dataloading/shadow.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
+"""ShaDow-GNN subgraph samplers."""
+from ..utils import prepare_tensor_or_dict
+from ..base import NID
+from ..sampling import sample_neighbors
+from .neighbor import NeighborSamplingMixin
+from .dataloader import exclude_edges, Sampler
+
+class ShaDowKHopSampler(NeighborSamplingMixin, Sampler):
+    """K-hop subgraph sampler used by
+    `ShaDow-GNN <https://openreview.net/forum?id=GIeGTl8EYx>`__.
```
Comment (mufeili) [inline python/dgl/dataloading/shadow.py:10]: It might be better to give the latest ArXiv link: https://arxiv.org/abs/2012.01380

File: python/dgl/dataloading/shadow.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        List of neighbors to sample per edge type for each GNN layer, starting from the
+        first layer.
+
+        If the graph is homogeneous, only an integer is needed for each layer.
+
+        If None is provided for one layer, all neighbors will be included regardless of
+        edge types.
+
+        If -1 is provided for one edge type on one layer, then all inbound edges
+        of that edge type will be included.
```
Comment (mufeili) [inline python/dgl/dataloading/shadow.py:30]: I feel the description for `fanouts` needs to be improved. 
1. Why not stick to `-1` or `None` for both cases?
2. Prepare two paragraphs, one for homogeneous graphs and one for heterogeneous graphs.
3. If `fanouts` is a `list[int]`, is `fanouts[0]` for the first GNN layer or the last GNN layer? Even if users have read the user guide, they may not clearly remember that. So it's better to spell it out.
4. Provide concrete examples below for the different cases mentioned here.

File: python/dgl/dataloading/shadow.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    replace : bool, default True
+        Whether to sample with replacement
+    prob : str, optional
```
Comment (mufeili) [inline python/dgl/dataloading/shadow.py:33]: I guess `prob` should be a tensor instead?

File: python/dgl/dataloading/shadow.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    To train a 3-layer GNN for node classification on a set of nodes ``train_nid`` on
+    a homogeneous graph where each node takes messages from 5, 10, 15 neighbors for
+    the first, second, and third layer respectively (assuming the backend is PyTorch):
+
+    >>> sampler = dgl.dataloading.ShaDowKHopSampler([5, 10, 15])
+    >>> dataloader = dgl.dataloading.NodeDataLoader(
+    ...     g, train_nid, sampler,
+    ...     batch_size=1024, shuffle=True, drop_last=False, num_workers=4)
+    >>> for input_nodes, output_nodes, (subgraph,) in dataloader:
+    ...     train_on(subgraph)
```
Comment (mufeili) [inline python/dgl/dataloading/shadow.py:48]: Print the values of `input_nodes`, `output_nodes` and `subgraph`.

File: python/dgl/dataloading/shadow.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+    If training on a heterogeneous graph and you want different number of neighbors for each
+    edge type, one should instead provide a list of dicts.  Each dict would specify the
```
Comment (mufeili) [inline python/dgl/dataloading/shadow.py:51]: Extra empty space before "Each"

File: python/dgl/dataloading/dataloader.py [hunk 7]
```diff
@@ -195,15 +204,67 @@ def _find_exclude_eids(g, exclude_mode, eids, **kwargs):
     else:
         raise ValueError('unsupported mode {}'.format(exclude_mode))
 
+class Sampler(object):
+    """An abstract class that takes in a graph and a set of seed nodes and returns a
+    structure representing a smaller portion of the graph for computation.  It can
```
Comment (mufeili) [inline python/dgl/dataloading/dataloader.py:209]: Extra empty space before "It"

File: python/dgl/dataloading/dataloader.py [hunk 8]
```diff
@@ -195,15 +204,67 @@ def _find_exclude_eids(g, exclude_mode, eids, **kwargs):
...
+    be either a list of bipartite graphs (i.e. :class:`BlockSampler`), or a single
+    subgraph (i.e. :class:`SubgraphSampler`).
```
Comment (mufeili) [inline python/dgl/dataloading/dataloader.py:211]: `SubgraphSampler` cannot be linked.

File: python/dgl/dataloading/dataloader.py [hunk 9]
```diff
@@ -195,15 +204,67 @@ def _find_exclude_eids(g, exclude_mode, eids, **kwargs):
...
+        raise NotImplementedError
+
+    def set_output_context(self, ctx):
+        """Set the device the generated block or subgraph will be output to.
+        This should only be set to a cuda device, when multi-processing is not
+        used in the dataloader (e.g., num_workers is 0).
+
+        Parameters
+        ----------
+        output_ctx : DGLContext, default None
```
Comment (mufeili) [inline python/dgl/dataloading/dataloader.py:248]: `output_ctx` -> `ctx`

File: python/dgl/dataloading/neighbor.py [hunk 10]
```diff
@@ -5,7 +5,55 @@
...
+    - :attr:`prob`: The edge feature name that stores the (unnormalized) probability.
+    - :attr:`fanouts`: The list of fanouts (either an integer or a dictionary of edge
+      types and integers).
+
+    The mixin will generate the following attributes:
+
+    - :attr:`prob_arrays`: List of DGL NDArrays containing the unnormalized probabilities
+      for every edge type.
+    - :attr:`fanout_arrays`: List of DGL NDArrays containing the fanouts for every edge
+      type at every later.
```
Comment (mufeili) [inline python/dgl/dataloading/neighbor.py:23]: `later` -> `layer`.

File: python/dgl/dataloading/neighbor.py [hunk 11]
```diff
@@ -5,7 +5,55 @@
...
+        elif self.prob_arrays is None:
+            # build prob_arrays only once
+            self.prob_arrays = [nd.array([], ctx=nd.cpu())] * len(g.etypes)
+
+    def _build_fanout(self, block_id, g):
+        assert not self.fanouts is None, \
+            "_build_fanout() should only be called when fanouts is not None"
+        # build fanout_arrays only once for each layer
+        while block_id >= len(self.fanout_arrays):
+            for i in range(len(self.fanouts)):
```
Comment (mufeili) [inline python/dgl/dataloading/neighbor.py:42]: L41 - L42 is a bit weird. Is it equivalent to use the minimum of `block_id` and `len(self.fanouts)`?

File: python/dgl/transform.py [hunk 12]
```diff
@@ -1303,8 +1304,8 @@ def laplacian_lambda_max(g):
         adj = g_i.adj(transpose=True, scipy_fmt=g_i.formats()['created'][0]).astype(float)
         norm = sparse.diags(F.asnumpy(g_i.in_degrees()).clip(1) ** -0.5, dtype=float)
         laplacian = sparse.eye(n) - norm * adj * norm
-        rst.append(sparse.linalg.eigs(laplacian, 1, which='LM',
-                                      return_eigenvectors=False)[0].real)
+        rst.append(scipy.sparse.linalg.eigs(
+            laplacian, 1, which='LM', return_eigenvectors=False)[0].real)
```
Comment (mufeili) [inline python/dgl/transform.py:1308]: What's the purpose of the modifications made in this file?

File: python/dgl/dataloading/neighbor.py [hunk 13]
```diff
@@ -30,6 +78,9 @@ class MultiLayerNeighborSampler(BlockSampler):
     return_eids : bool, default False
         Whether to return the edge IDs involved in message passing in the MFG.
         If True, the edge IDs will be stored as an edge feature named ``dgl.EID``.
+    prob : str, optional
+        If given, the probability of each neighbor being sampled is proportional
+        to the edge feature with the given name.
```
Comment (mufeili) [inline python/dgl/dataloading/neighbor.py:83]: Ditto

File: python/dgl/utils/internal.py [hunk 14]
```diff
@@ -691,6 +691,11 @@ def __getitem__(self, idx):
         g = self._groups[k]
         return k, g[j]
 
+def maybe_flatten_dict(data):
+    """Return a FlattenedDict if the input is a Mapping, or the data itself otherwise.
+    """
+    return FlattenedDict(data) if isinstance(data, Mapping) else data
```
Comment (mufeili) [inline python/dgl/utils/internal.py:697]: Is this used anywhere?

File: python/dgl/dataloading/neighbor.py (hunk 11 above)
Comment (BarclayII) [reply python/dgl/dataloading/neighbor.py:42]: I think it can be rewritten to `for i in range(len(self.fanouts))`.  @nv-dlasalle any idea why it's written this way?

File: python/dgl/transform.py (hunk 12 above)
Comment (BarclayII) [reply python/dgl/transform.py:1308]: To pass lint tests.  Otherwise lint will complain that the import `scipy.sparse.linalg` is never used.

File: python/dgl/dataloading/shadow.py (hunk 3 above)
Comment (BarclayII) [reply python/dgl/dataloading/shadow.py:31]: I simplified the interface so that there's only -1.  None is no longer accepted.  A minor breaking change though as I don't think people uses None explicitly instead of using `MultiLayerFullNeighborSampler`.

File: python/dgl/dataloading/shadow.py (hunk 4 above)
Comment (BarclayII) [reply python/dgl/dataloading/shadow.py:34]: This is how I put it now:
```
If given, the probability of each neighbor being sampled is proportional
to the edge feature value with the given name.  The feature must be
a scalar on each edge.
```

**Comment by mufeili** [comment]
Are you going to open follow-up PRs for adding/updating examples of SEAL, GraphSAINT and ShaDow-GNN?

File: docs/source/api/python/dgl.dataloading.rst [hunk 15]
```diff
@@ -39,6 +39,28 @@ the ``sample_blocks`` methods.
...
     :show-inheritance:
 
+Subgraph Iterators
+------------------
+Subgraph iterators iterate over the original graph in subgraphs. One should use subgraph
+iterators with ``GraphDataLoader`` like follows:
+
+.. code:: python
+
+   sgiter = dgl.dataloading.ClusterGCNSubgraphIterator(g, 100, '.', refresh=True)
```
Comment (mufeili) [inline docs/source/api/python/dgl.dataloading.rst:49]: `g, 100, '.', refresh=True` -> `g, num_partitions=100, cache_directory='.', refresh=True`

File: docs/source/api/python/dgl.dataloading.rst [hunk 16]
```diff
@@ -39,6 +39,28 @@ the ``sample_blocks`` methods.
...
+   dataloader = dgl.dataloading.GraphDataLoader(sgiter, batch_size=4, num_workers=0)
+   for subgraph_batch in dataloader:
+       train_on(subgraph_batch)
+
+.. autoclass:: dgl.dataloading.dataloader.SubgraphIterator
```
Comment (mufeili) [inline docs/source/api/python/dgl.dataloading.rst:54]: Will `.. autoclass::` display private methods like `_cache_file_path`, `_load_parts`, etc?

File: python/dgl/dataloading/shadow.py (hunk 4 above)
Comment (mufeili) [reply python/dgl/dataloading/shadow.py:34]: Add `in g.edata` after `the given name`. Also, are both `(E)` and `(E, 1)` acceptable? It might be good to spell it out in the doc.

**Review by mufeili — APPROVED** [review]
LGTM in general

File: python/dgl/dataloading/neighbor.py (hunk 11 above)
Comment (nv-dlasalle) [reply python/dgl/dataloading/neighbor.py:42]: I don't remember exactly it was written this way (it also doesn't appear entirely correct). Since we never partially build the fanout arrays, it works as is, but in theory if we built it partially, this wouldn't function properly. Especially now that it's part of a separate class, it probably should be something like to only build the  missing layers.
```
if len(self.fanouts) > len(self.fanout_arrays):
  for i in range(len(self.fanout_arrays), len(self.fanouts)):
```

**Comment by BarclayII** [comment]
They will be in separate PRs.

File: docs/source/api/python/dgl.dataloading.rst (hunk 16 above)
Comment (BarclayII) [reply docs/source/api/python/dgl.dataloading.rst:54]: No.

File: python/dgl/dataloading/neighbor.py (hunk 11 above)
Comment (BarclayII) [reply python/dgl/dataloading/neighbor.py:42]: Will fix in the next PR.
